### PR TITLE
relay log for oracle

### PR DESCRIPTION
**Pull request by cartersz**
### What problem does this PR solve?

Issue Number: close #xxx

### What is changed and how it works?
supporting  relay log feature

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: pkg/loader/translate.go [hunk 1]
```diff
@@ -119,3 +192,37 @@ func getDMLType(mut *pb.TableMutation) DMLType {
...
+	"smallint":    mysql.TypeShort,
+	"char":        mysql.TypeString,
+	"time":        mysql.TypeDuration,
+	"timestamp":   mysql.TypeTimestamp,
+	"tinyint":     mysql.TypeTiny,
+	"tinytext":    mysql.TypeTinyBlob,
+	"varchar":     mysql.TypeVarchar,
+	"var_string":  mysql.TypeVarString,
+	"year":        mysql.TypeYear,
+}
```
Comment (lichunzhu) [inline pkg/loader/translate.go:228]: TiDB parser has a method for this. We shouldn't rewrite another one.
https://github.com/pingcap/tidb/blob/master/parser/types/etc.go#L130

File: pkg/loader/translate.go [hunk 2]
```diff
@@ -57,15 +67,78 @@ func SecondaryBinlogToTxn(binlog *pb.Binlog) (*Txn, error) {
 	return txn, nil
 }
 
-func getColVals(table *pb.Table, cols []*pb.Column) (map[string]interface{}, error) {
+// SecondaryBinlogToOracleTxn translate the Binlog format into Oracle Txn
+func SecondaryBinlogToOracleTxn(binlog *pb.Binlog, tableRouter *router.Table, binlogFilter *bf.BinlogEvent) (*Txn, error) {
```
Comment (lichunzhu) [inline pkg/loader/translate.go:71]: suggest combining this function with the upper one. They are almost the same expect for the router logic.

File: drainer/relay.go [hunk 3]
```diff
@@ -65,11 +71,16 @@ func feedByRelayLogIfNeed(cfg *Config) error {
...
-func feedByRelayLog(r relay.Reader, ld loader.Loader, cp checkpoint.CheckPoint) error {
+func feedByRelayLog(r relay.Reader, ld loader.Loader, cp checkpoint.CheckPoint, cfg *Config) error {
 	checkpointTS := cp.TS()
 	lastSuccessTS := checkpointTS
 	r.Run()
 
+	tableRouter, binlogFilter, routeErr := genRouterAndBinlogEvent(cfg.SyncerCfg)
+	if routeErr != nil {
+		return errors.Annotate(routeErr, "when feed by relay log, gen router and filter failed")
+	}
```
Comment (lichunzhu) [inline drainer/relay.go:82]: parse router only for oracle type dsync.

File: drainer/relay_test.go [hunk 4]
```diff
@@ -100,7 +103,26 @@ func (s *relaySuite) TestFeedByRealyLog(c *check.C) {
...
+				User:                "user-1",
+				Password:            "password-1",
+				TLS:                 tls,
+				OracleServiceName:   "oracle-service-name-1",
+				OracleConnectString: "oracle-connect-string-1",
+				Table:               "new_checkpoint_table",
+			},
+		},
+	}
+	err = feedByRelayLog(reader, ld, cp, cfg)
```
Comment (lichunzhu) [inline drainer/relay_test.go:125]: Add another ut for mysql or tidb?

File: drainer/relay.go [hunk 5]
```diff
@@ -112,8 +120,18 @@ func feedByRelayLog(r relay.Reader, ld loader.Loader, cp checkpoint.CheckPoint)
...
+			var err error
+			var tableRouter *router.Table = nil
+			upperColName := false
+			if cfg.SyncerCfg.DestDBType == "oracle" {
+				upperColName = true
+				tableRouter, _, err = genRouterAndBinlogEvent(cfg.SyncerCfg)
+				if err != nil {
+					return errors.Annotate(err, "when feed by relay log, gen router and filter failed")
+				}
+			}
```
Comment (lichunzhu) [inline drainer/relay.go:133]: please put generating `tableRouter` outside this for loop

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM
@glorv PTAL

**Review by lichunzhu — APPROVED** [review]
LGTM

File: drainer/relay.go (hunk 5 above)
Comment (cartersz) [reply drainer/relay.go]: ok

**Review by glorv — APPROVED** [review]

**Pull request merged**